admin/modules/config/plugins.php

```diff
@@ -517,36 +680,53 @@
 
 	$plugins->run_hooks("admin_config_plugins_plugin_list");
 
+	$form = new Form('index.php?module=config-plugins&amp;action=upload', 'post', '', 1);
+	$table = new Table;
+	$table->construct_header($lang->upload_plugin_desc);
+	$table->construct_cell($form->generate_file_upload_box('plugin_upload', array('style' => 'width: 230px;')).
+	                       $form->generate_check_box('ignore_vers', 1, $lang->plugin_ignore_vers).
+	                       ' '.
+	                       $form->generate_submit_button($lang->install_uploaded_plugin));
+	$table->construct_row();
+	$table->output($lang->upload_plugin);
+	$form->end();
+
 	if(!empty($plugins_list))
 	{
 		$a_plugins = $i_plugins = array();
 
-		foreach($plugins_list as $plugin_file)
+		if(!empty($plugins_list))
 		{
-			require_once MYBB_ROOT."inc/plugins/".$plugin_file;
-			$codename = str_replace(".php", "", $plugin_file);
-			$infofunc = $codename."_info";
-
-			if(!function_exists($infofunc))
+			foreach($plugins_list as $plugin_file)
 			{
-				continue;
-			}
-
-			$plugininfo = $infofunc();
-			$plugininfo['codename'] = $codename;
+				$codename = str_replace('.php', '', $plugin_file);
+				$plugininfo = read_json_file(MYBB_ROOT."inc/plugins/{$codename}/plugin.json", $err_msg, /*$show_errs = */false);
+				if(!$plugininfo)
+				{
+					continue;
+				}
+				plugininfo_keys_to_raw($plugininfo, true);
+				$plugininfo['codename'] = $codename;
+				require_once MYBB_ROOT."inc/plugins/{$codename}/{$codename}.php";
+				$dyndescfunc = $codename.'_dyndesc';
+				if(function_exists($dyndescfunc))
+				{
+					$dyndescfunc($plugininfo['description']);
+				}
 
-			if(isset($active_plugins[$codename]))
-			{
-				// This is an active plugin
-				$plugininfo['is_active'] = 1;
+				if(isset($active_plugins[$codename]))
+				{
+					// This is an active plugin
+					$plugininfo['is_active'] = 1;
```

Comment:
abstracting the pluginInfo into a class with dedicated setters and strict return values would IMO be a better approach over an array with unknown indices if you are rewriting this part anyways atm (code completion in IDEs anyone?).